core/controller/src/main/scala/whisk/core/controller/API.scala

```diff
@@ -0,0 +1,277 @@
+import whisk.core.entity.WhiskAuthStore
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.entity._
+import whisk.core.entity.types._
+import whisk.core.entitlement._
+import whisk.core.entity.ActivationId.ActivationIdGenerator
+import whisk.core.loadBalancer.LoadBalancerService
+
+/**
+  * Abstract class which provides basic Directives which are used to construct route structures
+  * which are common to all versions of the Rest API.
+  */
+protected[controller] class SwaggerDocs(apipath: Uri.Path, doc: String)(implicit actorSystem: ActorSystem)
+    extends Directives {
+
+    /** Swagger end points. */
+    protected val swaggeruipath = "docs"
+    protected val swaggerdocpath = "api-docs"
+
+    def basepath(url: Uri.Path = apipath): String = {
+        (if (url.startsWithSlash) url else Uri.Path./(url)).toString
+    }
+
+    /**
+      * Defines the routes to serve the swagger docs.
+      */
+    val swaggerRoutes: Route = {
+        pathPrefix(swaggeruipath) {
+            getFromDirectory("/swagger-ui/")
+        } ~ path(swaggeruipath) {
+            redirect(s"$swaggeruipath/index.html?url=$apiDocsUrl", PermanentRedirect)
+        } ~ pathPrefix(swaggerdocpath) {
+            pathEndOrSingleSlash {
+                getFromResource(doc)
+            }
+        }
+    }
+
+    /** Forces add leading slash for swagger api-doc url rewrite to work. */
+    private def apiDocsUrl = basepath(apipath / swaggerdocpath)
+}
+
+protected[controller] object RestApiCommons {
+    def requiredProperties = Map(WhiskConfig.servicePort -> 8080.toString) ++
+            WhiskConfig.whiskVersion ++
+            WhiskAuthStore.requiredProperties ++
+            WhiskEntityStore.requiredProperties ++
+            WhiskActivationStore.requiredProperties ++
+            WhiskConfig.consulServer ++
+            EntitlementProvider.requiredProperties ++
+            WhiskActionsApi.requiredProperties ++
+            Authenticate.requiredProperties ++
+            Collection.requiredProperties
+}
+
+/**
+  * A trait for wrapping routes with headers to include in response.
+  * Useful for CORS.
+  */
+protected[controller] trait RespondWithHeaders extends Directives {
+    val allowOrigin = `Access-Control-Allow-Origin`.*
+    val allowHeaders = `Access-Control-Allow-Headers`("Authorization", "Content-Type")
+    val sendCorsHeaders = respondWithHeaders(allowOrigin, allowHeaders)
+}
+
+class API(config: WhiskConfig, apiPath: String, apiVersion: String)(
+    implicit val activeAckTopicIndex: InstanceId,
+    implicit val actorSystem: ActorSystem,
+    implicit val logging: Logging,
+    implicit val entityStore: EntityStore,
+    implicit val entitlementProvider: EntitlementProvider,
+    implicit val activationIdFactory: ActivationIdGenerator,
+    implicit val loadBalancer: LoadBalancerService,
+    implicit val activationStore: ActivationStore,
+    implicit val consulServer: String,
+    implicit val whiskConfig: WhiskConfig)
```

Comment:
Depends if this can be fixed or not:
```
ActionsApiTests.scala:72: ambiguous implicit values:
 both method executor in trait RouteTest of type => scala.concurrent.ExecutionContextExecutor
 and value executionContext in trait SequenceActions of type => scala.concurrent.ExecutionContextExecutor
 match expected type scala.concurrent.ExecutionContext
        waitOnView(entityStore, WhiskAction, namespace, 2)
```